### DISCUSSION TITLE
Confinement Factor

### DISCUSSION THREAD

**Discussion by smgitrep** [flexcompute/tidy3d#1786] [Q&A]
Does Tidy3D have a method to compute the confinement factor of waveguide modes in user specified regions?  Tidy3D does have a FluxMonitor, but it only applies to FDTD simulations.  Is there a similar monitor that can be used to compute the confinement factor from ModeSolverData?

**Comment by momchil-flex** [comment]
Could you specify what you mean by confinement factor? If you run the mode solver and then use `to_dataframe`, you can get various information about the modes, including mode area. Is this what you are referring to?

![image](https://github.com/flexcompute/tidy3d/assets/92756559/2621517f-bf93-4348-8f9e-ce1fe301162e)

**Reply by smgitrep** [reply]
What is the unit for mode area?  By confinement factor, I mean the ratio of the power flow through a 2D-region to the total power flow of a mode (which is normalized to unit power or 1).

**Reply by momchil-flex** [reply]
The unit of mode area is micron ** 2. The definition is as given e.g. [here](https://www.rp-photonics.com/effective_mode_area.html#:~:text=Definition%3A%20a%20quantitative%20measure%20of,covers%20in%20the%20transverse%20dimensions)

If you want to get the confinement factor the way you define it, I'm afraid we don't have a super explicit way to do it. However, I think you can do something like this to get the partial flux. Say you have `mode_data` computed already. `mode_data.dot(mode_data)` should be 1 as you say since we normalize to unit flux. Then you can do

```
# partial_size is the size of the plane over which you want to integrate
# make sure that one of the partial_size values is 0 as in the original mode solver (the normal direction)
monitor_partial = mode_data.monitor.updated_copy(size=partial_size) 
mode_data_partial = mode_data.updated_copy(monitor=monitor_partial)
partial_flux = mode_data_partial.dot(mode_data_partial)
```

Let me know if this works for you!

**Reply by smgitrep** [reply]
Thanks!  While it only returns the real part, this also works:

mode_data_partial.flux.isel(mode_index=0, f=0)

It returns a FluxDataArray, while the .dot() method returns a ModeAmpsDataArray.
